File: 2020/03/python/main.py
import math


def count_trees(lines: list[str], right: int, down: int) -> int:
    return len([line for i, line in enumerate(lines[::down])
                if line[i * right % len(line)] == '#'])


def solve(lines: list[str], slopes: list[tuple[int, int]]) -> int:
    return math.prod(count_trees(lines, r, d) for r, d in slopes)


def main(filename: str) -> (str, str):
    with open(filename, 'r') as input_file:
        input_lines = input_file.read().splitlines()
    solution1 = solve(input_lines, [(3, 1)])
    solution2 = solve(input_lines, [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)])
    return (str(solution1),
            str(solution2))


if __name__ == '__main__':
    solutions = main('../input.txt')
    print(solutions[0])
    print(solutions[1])

File: 2020/03/rust/src/main.rs
use std::error::Error;

fn main() -> Result<(), Box<dyn Error>>{
    let input = std::fs::read_to_string(r"..\input.txt")?;
    println!("{}", solve(&input, vec!((3, 1))));
    println!("{}", solve(&input, vec!((1, 1), (3, 1), (5, 1), (7, 1), (1, 2))));
    Ok(())
}

fn solve(input: &String, slopes: Vec<(usize, usize)>) -> usize{
    slopes.iter()
          .map(|&(right, down)| count_trees(&input, right, down))
          .fold(1, |p, x| p * x)
}

fn count_trees(input: &String, right: usize, down: usize) -> usize{
    input.lines()
         .step_by(down)
         .enumerate()
         .filter(|(i, row)| row.chars()
                               .nth(i * right % row.len()) == Some('#'))
         .count()
}
